### ISSUE TITLE
Add possibility to comment on lines of code in a pull request

### ISSUE THREAD

**Issue by hhenkel** [gogits/gogs#1644]
It would be nice to have the possibility to comment on specific lines of a code in a pull request - as it is possible in github:

![2015-09-15 13_51_58-added possibility to manage group members of a group by hhenkel pull request](https://cloud.githubusercontent.com/assets/2813782/9875814/824f1642-5bb1-11e5-92c2-cafb0ad7c471.png)

Gitbucket (another github clone written in scala) has it already implemented.

**Comment by Unknwon** [comment]
This is what I'm wanting, but take some time to clean up previous tasks. :smiley:

**Comment by bhchew** [comment]
+1

**Comment by nicolai86** [comment]
+1

**Comment by john-hu** [comment]
+1

**Comment by roidelapluie** [comment]
:+1:

**Comment by roidelapluie** [comment]
It would be great to be able to comment at the pull request level but also on individual commits.

**Comment by danielskowronski** [comment]
+1

**Comment by wsantos** [comment]
+1

**Comment by mgracer48** [comment]
+9000000000000000000000000000000000001

**Comment by mikehaertl** [comment]
+1 
For us this is one of the final main parts to make it basically feature complete. Having comments on code lines (and commits #796) is really a must. Would be a nice bonus for 0.10.0 ;)

**Comment by JonLundy** [comment]
+1

**Comment by redpist** [comment]
+1

**Comment by sergey-miryanov** [comment]
+1

**Comment by Chipsoid** [comment]
+1

**Comment by dhaval-aspire** [comment]
+1

**Comment by estan** [comment]
@Unknwon: Any update on this? We're thinking of switching from an SVN based workflow to Git + Gogs, but we'd really need this feature. We currently do code review using a very manual process (patches + e-mail), but even there we can comment on sections of code inline :) In fact, I would say most comments are inline.

Great work on Gogs btw!

**Comment by lostsnow** [comment]
+1 It is useful for review code in pull request

**Comment by ivanguajana** [comment]
+1

**Comment by toogy** [comment]
+1

**Comment by theredcat** [comment]
@ work We use gitlab but I'm looking for an alternative since their update and deployment process was written by crazy monkey in lack of caffeine.

But since pull request code line comments feature is missing we won't be able to switch to gogs since this is mandatory to do easy code review

**Comment by patrickreck** [comment]
+1

**Comment by strk** [comment]
May be possibly implemented via git-appraise ( #2210 )

**Comment by theredcat** [comment]
@strk As an ops, using the git-appraise cli would be fine for me. But some of my devs would like to have a nice UI to do this.

Frontend developpers, you know....

**Comment by strk** [comment]
@theredcat yes I understand that, what I'm saying is that Gogs could just add a GUI to the git-appraise tool. That way users have a choice to either use the commandline or the GUI, and peer reviews would go togheter with the code when repository is cloned.

**Comment by theredcat** [comment]
@strk Thx for the clarification. This would be, indeed, a very nice approach

**Comment by toogy** [comment]
:+1: `git-appraise` with a GUI sounds like the perfect road to follow

**Comment by stilet** [comment]
+1

**Comment by hhenkel** [comment]
@strk I think git-appraise would do the job but from what I recall (not sure if it was on gitter or in an issue discussion) @Unknwon wants to keep the way gogs handles the pull request stuff. From my understanding that does result in haveing this implemented for the gogs native way of handling pull request and if someone does the git-appraise way of storing data it has to be handled differently. If the later is done I expect that a user/admin has to choose which implementation of PRs he wants to use.

**Comment by javydreamercsw** [comment]
Agree, this is a must.

**Comment by nfx** [comment]
this is a must

**Comment by snoozeMoose** [comment]
+1

**Comment by videmort** [comment]
+9999999999

**Comment by MehSha** [comment]
+1

**Comment by eigan** [comment]
# STOP WITH '+1'

> In many cases, especially on popular projects, the result is a long thread full of emoji and not much content, which makes it difficult to have a discussion. With reactions, you can now reduce the noise in these threads.

https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments

**USE THAT!** 

I, and many others, are subscribed here for updates. Your '+1' is just making a mess.

**Comment by Unknwon** [comment]
Planned after https://github.com/gogits/gogs/issues/2109

**Comment by federicobond** [comment]
Hi! Any updates on this issue?

**Comment by snoozeMoose** [comment]
Nope. In general, the maintainer of gogs has a lot weird opinions of how to run the project (which is ok since it he owns it) so most people have abandoned gogs in favour of gitea.

**Comment by danielskowronski** [comment]
@Unknwon deleting comments? Seriously?...

> Nope. In general, the maintainer of gogs has a lot weird opinions of how to run the project (which is ok since it he owns it) so most people have abandoned gogs in favour of gitea.
~ @snoozeMoose

**Comment by OsmeDiegoGH** [comment]
Any updates??

**Comment by Armored-Dragon** [comment]
Any updates now? It's almost 2018, and this issue was opened in 2015.

**Comment by kathiouchka** [comment]
Hello - Will this feature be implemented at all?

**Comment by ekermawi** [comment]
@kathiouchka I would suggest moving to gitlab.
its now 3 years and they didn't do anything

**Comment by tgy** [comment]
There's a much more active fork of Gogs called [Gitea](https://github.com/go-gitea/gitea). I believe this feature hasn't been implemented there either but the plan is to use `git-appraise` to take care of it, which is a very nice solution.

**Comment by kathiouchka** [comment]
Thanks, the prob is my whole infra is based on gogs, everything is great, BUT this feature. Meh, i'll wait another year or two.

jk, hello gitlab!

**Comment by qRoC** [comment]
@tgy wait, author remove your messages about Gitea.

**Comment by Unknwon** [comment]
Choose GitLab if you need feature-rich Git hosting! **Seriously**! Or just pay GitHub, it's just $7/month. Or fork it, and do whatever you want!

**Comment by adventurist** [comment]
but a maintained Gogs with code comments is so much better than GitLab

**Comment by Freddis** [comment]
It's sucks when author writes things like above. Very unprofessional.
Take a look on your advertisement: "A painless self-hosted Git service."

Commenting on lines of code is a basic commodity for git hosting, not some extra feature. I wait for this change for 2.5 years, it's pretty sad, that you don't give a damn about your followers.

**Comment by Unknwon** [comment]
@Freddis oh wow, thanks for such a great statement! For over 2.5 years you finally concluded with a paragraph and not a single moment to fork?

I have zero interest on philosophy discussion, thus I'll delete both of our comments.

**Comment by forestbelton** [comment]
What needs to be done so that this can become a reality? I see that there may be a dependency on #2109, but I don't see any progress on that issue.

**Comment by ptman** [comment]
@forestbelton at the risk of getting this comment deleted, gitea has comments available in the 1.6.0 release

**Comment by minias** [comment]
+1

**Comment by minias** [comment]
Are you going to update it?

**Comment by lcalaresu** [comment]
I'm interested into this. Is anyone already wordking on it ?

**Comment by christopherstock** [comment]
Any updates on this major missing feature in Gogs? :D <3

**Comment by ivanbogomoloff** [comment]
i start made this feature, but my progress is slow
my progress on this video https://youtu.be/2Prxnibd8IM
branch here https://github.com/ivanbogomoloff/gogs/commits/feature/1644/ability-comment-pr  

nowaday i implemented:
- simple ui on html+js (add, delete, markdown render)
- http post endpoint with send comment data and response rendered code comment
but now it very raw

i plan finish this feature in very very simple implementation to august, i hope do it

**Comment by Aleksei-Badyaev** [comment]
The developers of Gogs have no plans at all to implement this great feature?

**Comment by timstableford** [comment]
Fyi, it is possible in Gitea now which is a fork of Gogs
